Utils/Event Pump.cpp

```diff
@@ -601,67 +626,172 @@ EV_S_GETNEWPATH				SGetNewPath;
 EV_S_BEGINTURN				SBeginTurn;
 EV_S_CHANGESTANCE			SChangeStance;
 EV_S_SETDIRECTION			SSetDirection;
-EV_S_SETDESIREDDIRECTION			SSetDesiredDirection;
-EV_S_BEGINFIREWEAPON	SBeginFireWeapon;
+EV_S_SETDESIREDDIRECTION	SSetDesiredDirection;
+EV_S_BEGINFIREWEAPON		SBeginFireWeapon;
 EV_S_FIREWEAPON				SFireWeapon;
 EV_S_WEAPONHIT				SWeaponHit;
 EV_S_STRUCTUREHIT			SStructureHit;
 EV_S_WINDOWHIT				SWindowHit;
-EV_S_MISS							SMiss;
-EV_S_NOISE						SNoise;
+EV_S_MISS					SMiss;
+EV_S_NOISE					SNoise;
 EV_S_STOP_MERC				SStopMerc;
-EV_S_SENDPATHTONETWORK SUpdateNetworkSoldier;
+EV_S_SENDPATHTONETWORK		SUpdateNetworkSoldier;
 
 extern	BOOLEAN				gfAmINetworked;
 
-BOOLEAN AddGameEventToQueue( UINT32 uiEvent, UINT16 usDelay, PTR pEventData, UINT8 ubQueueID );
+
 BOOLEAN ExecuteGameEvent( EVENT *pEvent );
 
 
-BOOLEAN AddGameEvent( UINT32 uiEvent, UINT16 usDelay, PTR pEventData )
+BOOLEAN AddEvent(UINT32 uiEvent, UINT16 usDelay, PTR pEventData, UINT32 uiDataSize, UINT8 ubQueueID)
 {
-	if (usDelay == DEMAND_EVENT_DELAY)
-	{
-		//DebugMsg( TOPIC_JA2, DBG_LEVEL_3, String("AddGameEvent: Sending Local and network #%d", uiEvent));
-		#ifdef NETWORKED
-		if(gfAmINetworked)
-			SendEventToNetwork(uiEvent, usDelay, pEventData);
-		#endif
-		return( AddGameEventToQueue( uiEvent, 0, pEventData, DEMAND_EVENT_QUEUE ) );
-	}
-	else if( uiEvent < EVENTS_LOCAL_AND_NETWORK)
-	{
-		//DebugMsg( TOPIC_JA2, DBG_LEVEL_3, String("AddGameEvent: Sending Local and network #%d", uiEvent));
-		#ifdef NETWORKED
-		if(gfAmINetworked)
-			SendEventToNetwork(uiEvent, usDelay, pEventData);
-		#endif
-		return( AddGameEventToQueue( uiEvent, usDelay, pEventData, PRIMARY_EVENT_QUEUE ) );
+	BYTE* pData = (BYTE*)MemAlloc(uiDataSize);
+	Assert(pData);
+	memcpy(pData, pEventData, uiDataSize);
+
+	// Add event to queue
+	switch (ubQueueID)
+	{
+	case PRIMARY_EVENT_QUEUE:
+		//hEventQueue.push(*pEvent);
+		hEventQueue.emplace(
+			EVENT{
+				GetJA2Clock(),	
+				0,				
+				usDelay,		
+				uiEvent,		
+				uiDataSize,		
+				pData			
+			}

```

Comment:
suggest you follow Gen. Melchott's code snippet here, this is way too much boilerplate that should be in EVENT's constructor